Add support for VERSION field type in SQL and EQL

luigidellaquila opened the pull request:
Fixes  #83375

luigidellaquila committed:
Add support for VERSION field type in SQL and EQL

luigidellaquila committed:
Update docs/changelog/85502.yaml

elasticsearchmachine commented:
Hi @luigidellaquila, I've created a changelog YAML for you.

luigidellaquila commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DataTypeConverter.java:
```diff
@@ -537,6 +548,10 @@ public static Object convert(Object value, DataType dataType) {
             return o;
+        }),
+        STRING_TO_VERSION(o -> {
+            // TODO validate...?
```
Not sure this is needed. The underlying implementation seems to do no validation on VERSION fields

luigidellaquila committed:
Add SQL csv-spec tests for VERSION fields

luigidellaquila committed:
Merge remote-tracking branch 'luigidellaquila/enhancement/support_version_data_type_in_ql' into enhancement/support_version_data_type_in_ql

luigidellaquila committed:
Fix test cases

luigidellaquila commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enhancement/support_version_data_type_in_ql

luigidellaquila committed:
Fix test to consider v 8.3 as version of introduction of the feature

luigidellaquila committed:
Fix frozen csv test cases

luigidellaquila commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enhancement/support_version_data_type_in_ql

luigidellaquila commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enhancement/support_version_data_type_in_ql

luigidellaquila committed:
Fix further tests

luigidellaquila commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enhancement/support_version_data_type_in_ql

luigidellaquila committed:
Merge remote-tracking branch 'luigidellaquila/enhancement/support_version_data_type_in_ql' into enhancement/support_version_data_type_in_ql

Luegg commented on x-pack/plugin/sql/qa/server/src/main/resources/version.csv-spec:
```diff
@@ -0,0 +1,69 @@
+
+//
+// Tests for VERSION fields
+//
```
a couple more things that I would cover:

* group by version
* constant expression returning a version (cast)
* select a string field casted to a version (script returning a version value). like `select concat(str, '.0')::version from "versions"`
* order by script returning a version
* group by script returning a version
* cast a version to a string

luigidellaquila replied:
Thank you @Luegg, it revealed a problem:

```
select '1.2.3'::version < '1.11.0'::version
```

the comparison takes into consideration SemVer order, we will have to reimplement it in *QL (now it's just comparing strings, that is wrong)

luigidellaquila replied:
Just for the record, the same happens with `ip` field type

Luegg replied:
oh, I didn't see your response. I also created an issue for the IP case https://github.com/elastic/elasticsearch/issues/85563

Luegg reviewed (commented):
looks mostly good to me but I left some suggestions for more testing

luigidellaquila commented:
@elasticmachine run elasticsearch-ci/packaging-tests-windows-sample

luigidellaquila committed:
Implement comparison of Version fields (folding)

luigidellaquila commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enhancement/support_version_data_type_in_ql

luigidellaquila committed:
Manage compatibility

and add new test cases

luigidellaquila committed:
Fix test cases

luigidellaquila commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enhancement/support_version_data_type_in_ql

luigidellaquila committed:
Add support for ORDER BY versionValueCalculatedWithScript

luigidellaquila committed:
Merge branch 'master' into enhancement/support_version_data_type_in_ql

astefan commented on x-pack/plugin/sql/qa/server/src/main/resources/version.csv-spec:
```diff
@@ -0,0 +1,206 @@
+// To mute tests follow example in file: example.csv-spec
```
I suggest adding more tests with queries with:
- `HAVING`
- `GROUP BY` and `COUNT`
- `NULLS FIRST`
- other functions using `version`, including in `ORDER BY`. For example `SELECT CASE WHEN LEFT(version, 1) = '1' THEN 1 WHEN LEFT(version, 1) = '2' THEN 2 ELSE 0 AS v, COUNT(*) GROUP BY 1 ORDER BY 1 DESC`.
- `ORDER BY` with `DESC`
- `SELECT version, version as v, version as v2 GROUP BY 1 ORDER BY 2 LIMIT 3`

luigidellaquila replied:
HAVING unfortunately is supported only for a limited set of data types, and Version is not one of them.

astefan commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/agg/GroupByKey.java:
```diff
@@ -62,6 +62,7 @@ public final CompositeValuesSourceBuilder<?> asValueSource() {
             } else if (script.outputType() == IP) {
                 builder.userValuetypeHint(ValueType.IP);
             }
+            // TODO hint for VERSION
```
?

luigidellaquila replied:
I see `ValueType` is  deprecated, so I was unsure if adding further content to that class would be acceptable and if there could be an alternative. Any suggestions?

luigidellaquila replied:
After off-line discussions and investigations, it seems it's not needed.

astefan commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/FieldAttributeTests.java:
```diff
@@ -182,7 +182,7 @@ public void testDottedFieldPathTypo() {
     public void testStarExpansionExcludesObjectAndUnsupportedTypes() {
```
I think you need to look at `testUnsignedLongVersionCompatibility` as a test example for `version`. What do you think @bpintea?

astefan commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plan/logical/command/ShowColumnsTests.java:
```diff
@@ -94,4 +97,16 @@ public void testUnsignedLongFiltering() {
+            // mapping's mutated by IndexCompatibility.compatible, needs to stay in the loop
+            Map<String, EsField> mapping = loadMapping("mapping-multi-field-variation.json", true);
+            ShowColumns.fillInRows(IndexCompatibility.compatible(mapping, Version.fromId(version.id)), null, rows);
+            assertTrue((supportsUnsignedLong(Version.fromId(version.id)) && rows.contains(rowSupported)) || rows.contains(rowUnsupported));
```
`supportsUnsignedLong`?

astefan reviewed (commented):
Great progress, but I think more tests are needed. Added some suggestions, feel free to check other already existent tests to think about other ideas, as well.

luigidellaquila commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/ResultSetTestCase.java:
```diff
@@ -2470,6 +2497,48 @@ private static void indexTestFieldsDoc(String docId, Object... values) throws IO
+        });
+
+        // some higher versions
+        for (int i = 0; i < randomInt(10); i++) {
```
It was an attempt to randomize the test dataset as much as possible, adding valid and invalid version values outside of the range of the test queries so that the tests are still predictable. Eg. test that random values that do not match the test criteria are not returned by mistake.

luigidellaquila committed:
Add further tests and address review comments

luigidellaquila committed:
Merge branch 'master' into enhancement/support_version_data_type_in_ql

luigidellaquila committed:
Fix based on review suggestions

astefan reviewed (approved):
LGTM. Thank you.

Luegg reviewed (commented):
Some additional comments also partially related to our offline discussion about extracting versions from search results as instances of `Version`.

luigidellaquila committed:
Merge branch 'master' into enhancement/support_version_data_type_in_ql

luigidellaquila commented on x-pack/plugin/sql/qa/server/src/main/resources/version.csv-spec:
```diff
@@ -0,0 +1,367 @@
+
+
+castConstantToVersion
+SELECT '1.2.3'::version as v;
```
I tried it, but it seems to work fine. Are you testing it on the same dataset or on something different?
Anyway, I'm adding a specific test case to the csv specs.

luigidellaquila commented on x-pack/plugin/sql/qa/server/src/main/resources/version.csv-spec:
```diff
@@ -0,0 +1,367 @@
+;
+
+orderByVersion
+SELECT * FROM apps ORDER BY version, id;
```
At first I thought it could depend on valid vs. invalid version numbers, but I tried it on different datasets and it seems to be completely independent from that.
I investigated it a bit more and it seems it depends on the same problem reported in #58820

luigidellaquila commented:
> There seems to be a problem with queries like select `first(version::string::version) from apps` though.

The reason for this error is that StreamOutput has no writer defined for Version type. The blocker for fixing it is that Version class is in x-pack, so I cannot reference it from inside Server module. Unfortunately I don't see a simple solution for this, apart from moving Version implementation to Server module, with all the implications on licensing. 
In my opinion, a fix with this impact (changes to multiple modules, licensing implications) is not justified by the uncommon use case, unless there is some request from the user base. My suggestion is to leave it as it is for now and track it with a new issue.

luigidellaquila committed:
Add further fixes based on reviews

- better type checks (avoid unpredictable automatic casts)
- support multiple casts
- add further test cases

luigidellaquila committed:
Merge branch 'master' into enhancement/support_version_data_type_in_ql

luigidellaquila committed:
Update version for bwc checks (to v 8.4.0)

luigidellaquila commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enhancement/support_version_data_type_in_ql

bpintea commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/FieldAttributeTests.java:
```diff
@@ -349,6 +351,45 @@ public void testUnsignedLongVersionCompatibility() {
         }
     }
 
+    public void testVersionFieldVersionCompatibility() {
```
`testVersionTypeVersionCompatibilty()`? `SELECT '1.2.3'::version'...` makes use of no field to test.

bpintea commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysColumnsTests.java:
```diff
@@ -170,6 +179,30 @@ public void testUnsignedLongFiltering() {
         }
     }
 
+    public void testVersionFieldFiltering() {
```
`testVersionTypeFiltering()`?

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DataTypes.java:
```diff
@@ -44,6 +44,8 @@ public final class DataTypes {
     // ip
     public static final DataType IP               = new DataType("ip",                45,                false, false, true);
+    // version
+    public static final DataType VERSION          = new DataType("version",           Integer.MAX_VALUE, false, false, true);
```
Giving VERSION a size of Integer.MAX_VALUE seems wrong, though I guess that's the theoretical max. Not sure if ES type enforces any limit? It seems there's only a recommendation, but no actual limit: https://semver.org/#does-semver-have-a-size-limit-on-the-version-string

luigidellaquila replied:
Technically, any string can be indexed as a Version, so the limit is the same as a KEYWORD.

bpintea reviewed (approved):
Nice work, LGTM!

luigidellaquila commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/ResultSetTestCase.java:
```diff
@@ -2244,6 +2245,30 @@ public void testResultSetConsumed() throws IOException {
+            assertFalse(results.next());
+        });
+
+        // bad version value
+        query = "SELECT name, version from test where version = 'foo'";
+        doWithQuery(query, results -> {
+            results.next();
+            assertEquals("version foo", results.getString("name"));
+            assertEquals("foo", results.getString("version"));
+            assertFalse(results.next());
+        });
+    }
```
Since there is this distinction between valid and invalid versions, I thought it would be good to have both in the result set test. Anyway, I agree that knowing the implementation details this test is probably not crucial. I'll refactor it a bit to make it shorter.

Luegg reviewed (approved):
👍

luigidellaquila commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/EsType.java:
```diff
@@ -32,6 +32,7 @@ public enum EsType implements SQLType {
     TIME(Types.TIME),
     DATETIME(Types.TIMESTAMP),
     IP(Types.VARCHAR),
+    VERSION(Types.VARCHAR),
```
Good point, fixed

luigidellaquila commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/gen/script/Scripts.java:
```diff
@@ -80,7 +80,12 @@ public static ScriptTemplate nullSafeFilter(ScriptTemplate script) {
 
     public static ScriptTemplate nullSafeSort(ScriptTemplate script) {
-        String methodName = script.outputType().isNumeric() ? "nullSafeSortNumeric" : "nullSafeSortString";
+        String methodName;
```
Just a leftover, fixed

luigidellaquila committed:
Implement review suggestions

luigidellaquila commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enhancement/support_version_data_type_in_ql

costin reviewed (changes requested):
Requesting changes regarding the impact on mapper-version project.

luigidellaquila committed:
Revert unnescessary changes to mapper-version

luigidellaquila committed:
Merge remote-tracking branch 'luigidellaquila/enhancement/support_version_data_type_in_ql' into enhancement/support_version_data_type_in_ql

luigidellaquila commented on x-pack/plugin/mapper-version/src/main/java/org/elasticsearch/xpack/versionfield/VersionEncoder.java:
```diff
@@ -35,7 +35,7 @@
  * </ul>
  */
-class VersionEncoder {
+public class VersionEncoder {
```
Good point, actually these changes are a leftover from the previous attempts to support Version without impacting on the Search/Painless implementation, but since we went with a more complete solution (see https://github.com/elastic/elasticsearch/pull/85990), we don't need them anymore.

luigidellaquila committed:
Implement review suggestions

luigidellaquila commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enhancement/support_version_data_type_in_ql

luigidellaquila committed:
Merge remote-tracking branch 'luigidellaquila/enhancement/support_version_data_type_in_ql' into enhancement/support_version_data_type_in_ql

luigidellaquila committed:
Fix compile problems after merge from master

costin reviewed (approved):
Thanks for reviewing the feedback. LGTM!

luigidellaquila committed:
Fix failing test case (due to refactoring on master)

elasticmachine committed:
Merge branch 'master' into enhancement/support_version_data_type_in_ql

luigidellaquila commented:
Had to revert due to the following:

```
va.lang.IllegalArgumentException: can not write type [class org.elasticsearch.xpack.versionfield.Version]
        at org.elasticsearch.server@8.4.0-SNAPSHOT/org.elasticsearch.common.io.stream.StreamOutput.writeGenericValue(StreamOutput.java:828)
        at org.elasticsearch.xpack.ql.expression.gen.processor.ConstantProcessor.writeTo(ConstantProcessor.java:68)
        at org.elasticsearch.server@8.4.0-SNAPSHOT/org.elasticsearch.common.io.stream.StreamOutput.writeNamedWriteable(StreamOutput.java:1097)
        at org.elasticsearch.server@8.4.0-SNAPSHOT/org.elasticsearch.common.io.stream.StreamOutput.writeNamedWriteableList(StreamOutput.java:1203)
        at org.elasticsearch.xpack.sql.expression.predicate.conditional.ConditionalProcessor.writeTo(ConditionalProcessor.java:73)
        at org.elasticsearch.server@8.4.0-SNAPSHOT/org.elasticsearch.common.io.stream.StreamOutput.writeNamedWriteable(StreamOutput.java:1097)
        at org.elasticsearch.xpack.ql.execution.search.extractor.ComputingExtractor.writeTo(ComputingExtractor.java:56)
        at org.elasticsearch.server@8.4.0-SNAPSHOT/org.elasticsearch.common.io.stream.StreamOutput.writeNamedWriteable(StreamOutput.java:1097)
        at org.elasticsearch.server@8.4.0-SNAPSHOT/org.elasticsearch.common.io.stream.StreamOutput.writeNamedWriteableList(StreamOutput.java:1203)
        at org.elasticsearch.xpack.sql.execution.search.SearchHitCursor.writeTo(SearchHitCursor.java:80)
        at org.elasticsearch.server@8.4.0-SNAPSHOT/org.elasticsearch.common.io.stream.StreamOutput.writeNamedWriteable(StreamOutput.java:1097)
        at org.elasticsearch.xpack.sql.session.Cursors.encodeToString(Cursors.java:82)
        at org.elasticsearch.xpack.sql.session.Cursors.encodeToString(Cursors.java:73)
        at org.elasticsearch.xpack.sql.plugin.TransportSqlQueryAction.createResponse(TransportSqlQueryAction.java:225)
```

The problem seems to be related to how Version objects are translated to strings to be serialized, so it should be deterministic, but strangely it's not (before merging the build was green and I didn't notice this error before during the tests)
